### PR TITLE
cephfs-mirror: use monotonic clocks

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/62265

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
jenkins test signed

**Comment by joscollin** [comment]
jenkins retest this please

**Review by dparmar18 — APPROVED** [review]

**Review by rishabh-d-dave — APPROVED** [review]
Looks good otherwise.

File: src/tools/cephfs_mirror/Mirror.cc [hunk 1]
```diff
@@ -547,19 +547,18 @@ void Mirror::peer_removed(const Filesystem &filesystem, const Peer &peer) {
 void Mirror::update_fs_mirrors() {
   dout(20) << dendl;
 
-  auto now = ceph_clock_now();
-  double blocklist_interval = g_ceph_context->_conf.get_val<std::chrono::seconds>
-    ("cephfs_mirror_restart_mirror_on_blocklist_interval").count();
-  double failed_interval = g_ceph_context->_conf.get_val<std::chrono::seconds>
-    ("cephfs_mirror_restart_mirror_on_failure_interval").count();
+  std::chrono::seconds blocklist_interval = g_ceph_context->_conf.get_val<std::chrono::seconds>
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/tools/cephfs_mirror/Mirror.cc:550]: +1

[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/Mirror.cc:550]: Please note this is within `std`. what shorthand do you suggest ?

[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/Mirror.cc]: updated

File: src/tools/cephfs_mirror/Mirror.cc [hunk 2]
```diff
@@ -547,19 +547,18 @@ void Mirror::peer_removed(const Filesystem &filesystem, const Peer &peer) {
...
+    ("cephfs_mirror_restart_mirror_on_blocklist_interval");
+  std::chrono::seconds failed_interval = g_ceph_context->_conf.get_val<std::chrono::seconds>
+    ("cephfs_mirror_restart_mirror_on_failure_interval");
 
   {
     std::scoped_lock locker(m_lock);
     for (auto &[filesystem, mirror_action] : m_mirror_actions) {
       auto failed_restart = mirror_action.fs_mirror && mirror_action.fs_mirror->is_failed() &&
-	(failed_interval > 0 && (mirror_action.fs_mirror->get_failed_ts() - now) > failed_interval);
+	(failed_interval.count() > 0 && std::chrono::duration_cast<std::chrono::seconds>(mirror_action.fs_mirror->get_failed_ts() - clock::now()) > failed_interval);
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/Mirror.cc]: updated

**Comment by joscollin** [comment]
@rishabh-d-dave @dparmar18 Please check again. Updated the review comments as a separate commit.

**Comment by dparmar18** [comment]
> @rishabh-d-dave @dparmar18 Please check again. Updated the review comments as a separate commit.

LGTM :)

**Comment by joscollin** [comment]
squashed commits

**Comment by joscollin** [comment]
> @joscollin Since you pushed another set of changes after Dhairya approved, is this PR ready? If so, I'll pick it up for QA.

@rishabh-d-dave Please pick up for qa. Thanks.

**Review by rishabh-d-dave — APPROVED** [review]
I had to do multiple QA runs and re-runs due to build issues (due to one of the PRs in testing branch), unrelated issues in FS test suite, unrelated issues in testing kernel and new failures in every run that needed to ensured that they were unrelated to PRs in testing branch.

I finally have a run that is better than others - https://tracker.ceph.com/projects/cephfs/wiki/Main#4-Apr-2024. Sorry for the delay.

**Pull request merged**